### PR TITLE
feat: add support for varargs

### PR DESCRIPTION
**Pull request by chanattan**
WIP.
Related to #8158.

### CODE REVIEW

**Comment by mlutze** [comment]
Since it's marked as a draft, I will hold off on reviewing, but let me know if I should take a look.

**Comment by magnus-madsen** [comment]
What's needed to push this over the finish line? :)

**Comment by chanattan** [comment]
I am still testing some cases on my own, it should be half-way done.

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 1]
```diff
@@ -222,6 +239,35 @@ object Simplifier {
       throw InternalCompilerException(s"Unexpected expression: $exp0.", loc)
   }
 
+  /**
+   * Helper method to get a corresponding monotype to the given type.
+   */
+  private def getMonotype(tpe: Type): MonoType =
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:245]: Can we do `visitType` for this one?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 2]
```diff
@@ -258,22 +279,53 @@ object TypeReduction {
    * @param methodName the potential candidate method's name
    * @param ts         the list of parameter types of the potential candidate method
    */
-  private def isCandidateMethod(cand: Method, methodName: String, isStatic: Boolean = false, ts: List[Type])(implicit flix: Flix): Boolean = {
+  private def isCandidateMethod(cand: Method, methodName: String, isStatic: Boolean = false, isVariadic: Boolean = false, ts: List[Type])(implicit flix: Flix): Boolean = {
     val static = java.lang.reflect.Modifier.isStatic(cand.getModifiers)
     (if (isStatic) static else !static) &&
       (cand.getName == methodName) &&
-      (cand.getParameterCount == ts.length) &&
+      // Either check candidate method if it is not variadic
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:286]: This is very hard to read. Can we refactor it into a series of guards?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 3]
```diff
@@ -258,22 +279,53 @@ object TypeReduction {
...
+      if (argsDiff <= 1) { // We remove the vararg array and check for required arguments
+        // Check subtyping for required args
+        // if (argsDiff <= 0) TODO INTEROP: support where argsDiff <= 0
+        (cand.getParameterTypes zip ts.slice(0, argsDiff + 1)).forall {
+          case (clazz, tpe) => isSubtype(tpe, Type.getFlixType(clazz))
+        } &&
+        // Check subtyping for varargs
+        ts.slice(argsDiff + 1, ts.length).forall(t => isSubtype(t, Type.getFlixType(tpe)))
+      } else false
+    } else false
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:326]: I would reverse these to be early-return guards. E.g. 
```
(if !cand.isVarArgs) {
  return false
}
```

File: main/src/ca/uwaterloo/flix/language/phase/Verifier.scala [hunk 4]
```diff
@@ -616,6 +616,49 @@ object Verifier {
     }
   }
 
+  /**
+   * Returns a corresponding Java class for the given monotype.
+   */
+  def getJavaType(tpe: MonoType, loc: SourceLocation): Class[_] = {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Verifier.scala:622]: Jonathan says this is fine

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 5]
```diff
@@ -220,10 +221,30 @@ object TypeReduction {
...
-      case 0 => JavaMethodResolutionResult.MethodNotFound
+      case 0 =>
+        // If no method was found, check if it is potentially a variadic method
+        val varMethods = clazz.getMethods.filter(m => isCandidateMethod(m, methodName, isStatic, isVariadic = true, ts))
+        varMethods.length match {
+          case 0 => JavaMethodResolutionResult.MethodNotFound
+          case 1 =>
+            val tpe = Type.Cst(TypeConstructor.JvmMethod(varMethods.head), loc)
+            JavaMethodResolutionResult.Resolved(tpe)
+          case 2 =>
```
[the remark this answers is not in the document]
Comment (chanattan) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:235]: Yes thank you

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala (hunk 1 above)
Comment (chanattan) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:245]: Oh it seems right. I removed the function and tried with visitType instead.

File: main/src/ca/uwaterloo/flix/language/phase/Verifier.scala (hunk 4 above)
Comment (chanattan) [reply main/src/ca/uwaterloo/flix/language/phase/Verifier.scala:622]: Ok thank you. It isn't complete for now as we want to have a working version.

**Comment by chanattan** [comment]
The base should work, now `checkJavaSubtype` instead of `checkEq` seems indeed weaker but stronger in that it requires a Java (class) comparison. That is why we have so many errors, we should probably have two cases to match the two functions depending on the array element type. There we consider again a certain comparison between MonoTypes and Java types, what is your opinion @mlutze?

**Comment by mlutze** [comment]
> The base should work, now `checkJavaSubtype` instead of `checkEq` seems indeed weaker but stronger in that it requires a Java (class) comparison. That is why we have so many errors, we should probably have two cases to match the two functions depending on the array element type. There we consider again a certain comparison between MonoTypes and Java types, what is your opinion @mlutze?

Yeah that's tricky. Ugh.

Maybe instead of converting to a Java class, we can just create a new `isSubType` function that mostly does the same as `isJavaSubtype` but also allows Flix stuff. Essentially the same as the subtyping check you had in another phase.

**Comment by chanattan** [comment]
I see, I will think about this and implement it a bit later. Thank you!

**Comment by magnus-madsen** [comment]
We can skip the verifier check if it is varargs.

**Comment by chanattan** [comment]
> We can skip the verifier check if it is varargs.

May we add a flag to ArrayLit or create an equivalent operation so that it gets ignored? I believe I should still build something in the Simplifier but it should only get ignored in Verifier.

**Comment by mlutze** [comment]
> > We can skip the verifier check if it is varargs.
> 
> May we add a flag to ArrayLit or create an equivalent operation so that it gets ignored? I believe I should still build something in the Simplifier but it should only get ignored in Verifier.

Right. We can still do the Simplifier translation, but in Verifier, we can just do something on the `invoke` expression where we check if the method is varargs, and if it is, then we skip checking it.

Recall that Verifier is only there for our (the compiler developers) debugging purposes. In general, if it becomes a burden, we shouldn't bother unless we know there is a bug.

**Comment by magnus-madsen** [comment]
> Recall that Verifier is only there for our (the compiler developers) debugging purposes. In general, if it becomes a burden, we shouldn't bother unless we know there is a bug.

Exactly. It is only to help ourselves.

**Comment by chanattan** [comment]
> > > We can skip the verifier check if it is varargs.
> > 
> > 
> > May we add a flag to ArrayLit or create an equivalent operation so that it gets ignored? I believe I should still build something in the Simplifier but it should only get ignored in Verifier.
> 
> Right. We can still do the Simplifier translation, but in Verifier, we can just do something on the `invoke` expression where we check if the method is varargs, and if it is, then we skip checking it.
> 
> Recall that Verifier is only there for our (the compiler developers) debugging purposes. In general, if it becomes a burden, we shouldn't bother unless we know there is a bug.

I see thank you.
However the Simplifier transforms Invoke into an ApplyAtomic where we lose the method, as the arguments are evaluated before the handler expression. The only thing I see as a workaround is either a new AtomicOp (VarArray) or a flag in ArrayLit.

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 6]
```diff
@@ -258,21 +278,54 @@ object TypeReduction {
...
+  private def isCandidateVariadicMethod(cand: Method, ts: List[Type])(implicit flix: Flix): Boolean = {
+    if (!cand.isVarArgs)
+      return false
+    val paramCount = cand.getParameterCount
+    val varArr = cand.getParameterTypes.last
+    val tpe = varArr.getComponentType
+    // Check minimum required arity
+    // e.g., for a defined function f(T arg1, T... args) arg1 is required.
+    val argsDiff = paramCount - ts.length
+    if (argsDiff > 1) // We remove the vararg array and check for required arguments
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:319]: you write that you now check something, but it just returns false, which makes the other method return false. is the comment outdated?

**Review by JonathanStarup — COMMENTED** [review]
Could you add some tests also? otherwise it seems good :)

**Comment by magnus-madsen** [comment]
Outdated. Will be fixed in separate PR.

**Pull request closed without merging**